File: go-client/internal/protocol/messages.go
// Package protocol provides message types, handler dispatch, file operations,
// and key migration logic.
package protocol

import (
	"time"

	"github.com/cisc468/p2p-project/internal/crypto"
)

const (
	ProtocolVersion = "1"
	ChunkSize       = 64 * 1024 // 64 KB
)

// Error codes
const (
	AuthFailed       = "AUTH_FAILED"
	UnknownPeer      = "UNKNOWN_PEER"
	FileNotFound     = "FILE_NOT_FOUND"
	ConsentDenied    = "CONSENT_DENIED"
	IntegrityFailure = "INTEGRITY_FAILURE"
	MigrationInvalid = "MIGRATION_INVALID"
	ProtocolErrorCode = "PROTOCOL_ERROR"
	VersionMismatch  = "VERSION_MISMATCH"
)

func envelope(msgType string, seq int, extra map[string]interface{}) map[string]interface{} {
	msg := map[string]interface{}{
		"type":      msgType,
		"version":   ProtocolVersion,
		"seq":       float64(seq),
		"timestamp": float64(time.Now().UnixMilli()),
	}
	for k, v := range extra {
		msg[k] = v
	}
	return msg
}

func ListRequest(seq int, ownerID *string) map[string]interface{} {
	var oid interface{} = nil
	if ownerID != nil {
		oid = *ownerID
	}
	return envelope("LIST_REQUEST", seq, map[string]interface{}{
		"filter": map[string]interface{}{"owner_id": oid},
	})
}

func ListResponse(seq int, files []map[string]interface{}) map[string]interface{} {
	// Convert nil slice to empty array for JSON
	if files == nil {
		files = []map[string]interface{}{}
	}
	ifiles := make([]interface{}, len(files))
	for i, f := range files {
		ifiles[i] = f
	}
	return envelope("LIST_RESPONSE", seq, map[string]interface{}{
		"files": ifiles,
	})
}

func TransferRequest(seq int, fileID, requesterID string, manifest map[string]interface{}) map[string]interface{} {
	extra := map[string]interface{}{
		"file_id":      fileID,
		"requester_id": requesterID,
	}
	if manifest != nil {
		extra["manifest"] = manifest
	}
	return envelope("TRANSFER_REQUEST", seq, extra)
}

func TransferResponse(seq int, fileID string, accepted bool, reason string) map[string]interface{} {
	return envelope("TRANSFER_RESPONSE", seq, map[string]interface{}{
		"file_id":  fileID,
		"accepted": accepted,
		"reason":   reason,
	})
}

func TransferChunk(seq int, fileID string, chunkIndex, totalChunks int, chunkData []byte) map[string]interface{} {
	return envelope("TRANSFER_CHUNK", seq, map[string]interface{}{
		"file_id":      fileID,
		"chunk_index":  float64(chunkIndex),
		"total_chunks": float64(totalChunks),
		"chunk_data":   crypto.B64URLEncode(chunkData),
	})
}

func TransferComplete(seq int, fileID, sha256Hash string) map[string]interface{} {
	return envelope("TRANSFER_COMPLETE", seq, map[string]interface{}{
		"file_id": fileID,
		"sha256":  sha256Hash,
	})
}

func KeyMigration(seq int, oldPeerID, newPeerID string,
	effectiveTS, expiryTS int64, reason, oldSig, newSig string) map[string]interface{} {
	return envelope("KEY_MIGRATION", seq, map[string]interface{}{
		"old_peer_id":         oldPeerID,
		"new_peer_id":         newPeerID,
		"effective_timestamp": float64(effectiveTS),
		"expiry_timestamp":    float64(expiryTS),
		"reason":              reason,
		"old_signature":       oldSig,
		"new_signature":       newSig,
	})
}

func MigrationAck(seq int, accepted bool, reason string) map[string]interface{} {
	return envelope("MIGRATION_ACK", seq, map[string]interface{}{
		"accepted": accepted,
		"reason":   reason,
	})
}

func ErrorMsg(seq int, code, message string) map[string]interface{} {
	return envelope("ERROR", seq, map[string]interface{}{
		"code":    code,
		"message": message,
	})
}

File: python-client/p2pshare/protocol/messages.py
"""Protocol message types and construction helpers.

All message types share a common envelope:
  { "type": ..., "version": "1", "seq": N, "timestamp": ms }
"""

import time
import base64

PROTOCOL_VERSION = "1"
CHUNK_SIZE = 64 * 1024  # 64 KB


def _envelope(msg_type: str, seq: int, **kwargs) -> dict:
    msg = {
        "type": msg_type,
        "version": PROTOCOL_VERSION,
        "seq": seq,
        "timestamp": int(time.time() * 1000),
    }
    msg.update(kwargs)
    return msg


def list_request(seq: int, owner_id: str | None = None) -> dict:
    return _envelope("LIST_REQUEST", seq, filter={"owner_id": owner_id})


def list_response(seq: int, files: list[dict]) -> dict:
    return _envelope("LIST_RESPONSE", seq, files=files)


def transfer_request(seq: int, file_id: str, requester_id: str, manifest: dict | None = None) -> dict:
    extra: dict = {"file_id": file_id, "requester_id": requester_id}
    if manifest is not None:
        extra["manifest"] = manifest
    return _envelope("TRANSFER_REQUEST", seq, **extra)


def transfer_response(seq: int, file_id: str, accepted: bool, reason: str = "") -> dict:
    return _envelope("TRANSFER_RESPONSE", seq, file_id=file_id, accepted=accepted, reason=reason)


def transfer_chunk(seq: int, file_id: str, chunk_index: int, total_chunks: int, chunk_data: bytes) -> dict:
    from p2pshare.crypto.keys import b64url_encode
    return _envelope(
        "TRANSFER_CHUNK", seq,
        file_id=file_id,
        chunk_index=chunk_index,
        total_chunks=total_chunks,
        chunk_data=b64url_encode(chunk_data),
    )


def transfer_complete(seq: int, file_id: str, sha256_hash: str) -> dict:
    return _envelope("TRANSFER_COMPLETE", seq, file_id=file_id, sha256=sha256_hash)


def key_migration(seq: int, old_peer_id: str, new_peer_id: str,
                  effective_ts: int, expiry_ts: int, reason: str,
                  old_signature: str, new_signature: str) -> dict:
    return _envelope(
        "KEY_MIGRATION", seq,
        old_peer_id=old_peer_id,
        new_peer_id=new_peer_id,
        effective_timestamp=effective_ts,
        expiry_timestamp=expiry_ts,
        reason=reason,
        old_signature=old_signature,
        new_signature=new_signature,
    )


def migration_ack(seq: int, accepted: bool, reason: str = "") -> dict:
    return _envelope("MIGRATION_ACK", seq, accepted=accepted, reason=reason)


def error_msg(seq: int, code: str, message: str) -> dict:
    return _envelope("ERROR", seq, code=code, message=message)


# Error codes
AUTH_FAILED = "AUTH_FAILED"
UNKNOWN_PEER = "UNKNOWN_PEER"
FILE_NOT_FOUND = "FILE_NOT_FOUND"
CONSENT_DENIED = "CONSENT_DENIED"
INTEGRITY_FAILURE = "INTEGRITY_FAILURE"
MIGRATION_INVALID = "MIGRATION_INVALID"
PROTOCOL_ERROR = "PROTOCOL_ERROR"
VERSION_MISMATCH = "VERSION_MISMATCH"
